src/Forum/Auth/SsoResponse.php

```diff
@@ -0,0 +1,175 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Forum\Auth;
+
+class SsoResponse
+{
+    protected $provider;
+
+    protected $identifier;
+
+    /**
+     * @var array
+     */
+    protected $provided = [];
+
+    /**
+     * @var array
+     */
+    protected $suggested = [];
+
+    /**
+     * @var mixed
+     */
+    protected $payload = [];
+
+    /**
+     * @var string
+     */
+    public function __construct(string $provider)
+    {
+        $this->provider = $provider;
+    }
+
+    /**
+     * @return string
+     */
+    public function getProvider(): string
+    {
+        return $this->provider;
+    }
+
+    /**
+     * @return string
+     */
+    public function getIdentifier(): string
+    {
+        return $this->identifier;
+    }
+
+    /**
+     * @param string $identifier
+     * @return $this
+     */
+    public function withIdentifier(string $identifier): self
+    {
+        $this->identifier = $identifier;
+
+        return $this;
+    }
+
+    /**
+     * @return array
+     */
+    public function getProvided(): array
+    {
+        return $this->provided;
+    }
+
+    /**
+     * @return array
+     */
+    public function getSuggested(): array
+    {
+        return $this->suggested;
+    }
+
+    /**
+     * @return mixed
+     */
+    public function getPayload()
+    {
+        return $this->payload;
+    }
+
+    /**
+     * @param string $key
+     * @param mixed $value
+     * @return $this
+     */
+    public function provide(string $key, $value): self
+    {
+        $this->provided[$key] = $value;
+
+        return $this;
+    }
+
+    /**
+     * @param string $email
+     * @return $this
+     *
+     * @deprecated in favor of provideEmail(), as email trustiness is now determined in admin settings.
+     */
+    public function provideTrustedEmail(string $email): self
+    {
+        return $this->provide('email', $email);
+    }
+
+    /**
+     * @param string $email
+     * @return $this
+     */
+    public function provideEmail(string $email): self
+    {
+        return $this->provide('email', $email);
+    }
+
+    /**
+     * @param string $url
+     * @return $this
+     */
+    public function provideAvatar(string $url): self
+    {
+        return $this->provide('avatar_url', $url);
+    }
+
+    /**
+     * @param string $key
+     * @param mixed $value
+     * @return $this
+     */
+    public function suggest(string $key, $value): self
+    {
+        $this->suggested[$key] = $value;
+
+        return $this;
+    }
+
+    /**
+     * @param string $username
+     * @return $this
+     */
+    public function suggestUsername(string $username): self
+    {
+        $username = preg_replace('/[^a-z0-9-_]/i', '', $username);
```

Comment:
How are we going to make this extendable ?

Can/should we re-use the regex from the UserValidator ?

Reply:
I would be fine running the username itself through the user validator, but the question is, what to do if it doesn't validate? Right now I'm thinking we just return an empty string, but should there be some notification/error message? Personally don't think so since it's suggested, but open to other ideas.

Reply:
Yeah that was just an idea, we could extract the regex from the validation rules, then use that regex like you did to filter out characters.

But then if an extension changes the validator rules and no longer uses a regex, this will break :thinking: